File: src/render.ts
import { palette as buildPalette, type Palette, type Variant } from "./color";
import { superellipse } from "./shape";
import { traits, type Traits } from "./traits";

export interface AvatarOptions {
  /** Which look to render. Default `"blob"`. Ignored by the per-variant entries. */
  variant?: Variant;
  /** Emits width/height attributes. Omit to let CSS size it (the viewBox always scales). */
  size?: number;
  /** Overrides the variant's own default. `false` renders transparent. */
  background?: boolean | "square" | "circle" | "squircle";
  /** Overrides specific palette entries. Overridden colors bypass the contrast guarantee. */
  palette?: Palette;
  /** Locks the hue in degrees, so the seed drives shape only. */
  hue?: number;
  /** Locks the `blob` tone as a 0–1 position in the swatch set. */
  tone?: number;
  /** Applies NFC + trim + lowercase to the seed. Default true. */
  normalize?: boolean;
  /** Enforces the minimum contrast ratios. Default true. */
  contrast?: boolean;
  /** Adds a <title> for screen readers. */
  title?: string;
}

export interface Style<L> {
  layout(t: Traits): L;
  render(l: L, p: Palette): string;
  background: boolean | "square" | "circle" | "squircle";
}

const escape = (s: string) =>
  s.replace(/[&<>]/g, c => (c === "&" ? "&amp;" : c === "<" ? "&lt;" : "&gt;"));

export function resolve<L>(style: Style<L>, variant: Variant, seed: string, opts: AvatarOptions) {
  const t = traits(seed, opts.normalize ?? true);
  return {
    t,
    palette: {
      ...buildPalette(
        opts.hue ?? t.num("hue", 0, 360),
        variant,
        opts.contrast ?? true,
        opts.tone ?? t("tone"),
      ),
      ...opts.palette,
    } as Palette,
  };
}

/**
 * Binds one style into an `avatar(seed, opts)` function.
 *
 * The wrapper exists so the per-variant entry points (`morphatar/blob`,
 * `morphatar/character`) can each pull in exactly one style. Importing the
 * dispatcher instead costs both, which the size gate will tell you about.
 */
export function makeAvatar<L>(style: Style<L>, variant: Variant) {
  return (seed: string, opts: AvatarOptions = {}): string => {
    const { t, palette: p } = resolve(style, variant, seed, opts);
    const out: string[] = [];

    // The backdrop is a shared concern, so variants declare a default rather
    // than each drawing their own.
    const bg = opts.background ?? style.background;
    if (bg === "square") {
      out.push(`<path d="M0 0H100V100H0Z" fill="${p.bg}"/>`);
    } else if (bg !== false) {
      const d = superellipse({ cx: 50, cy: 50, rx: 50, ry: 50, n: bg === "circle" ? 2 : 6 });
      out.push(`<path d="${d}" fill="${p.bg}"/>`);
    }

    out.push(style.render(style.layout(t), p));

    const dim = opts.size ? ` width="${opts.size}" height="${opts.size}"` : "";
    const title = opts.title ? `<title>${escape(opts.title)}</title>` : "";
    return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"${dim}>${title}${out.join("")}</svg>`;
  };
}

File: packages/flutter/lib/src/render.dart
/// Option resolution and the renderer seam.
///
/// Dart port of the static half of `packages/blobatar/src/render.ts` at
/// blobatar 2.4.0 (the pose/motion halves arrive in later port phases).
library;

import 'color.dart';
import 'expression.dart';
import 'shape.dart';
import 'traits.dart';
import 'styles/blob.dart' show style;
import 'styles/compose.dart';

/// How the backdrop is drawn. Mirrors the JS option values
/// `false | "square" | "circle" | "squircle"`.
///
/// The JS option also accepts `true`, which renders the squircle plate
/// (`n: 6`); use [Backdrop.squircle] for that case.
enum Backdrop { none, square, circle, squircle }

/// The options the layout and palette resolution read.
///
/// Animation belongs to a later port phase; static expressions are applied by
/// both the pure renderer and the Flutter painter.
class BlobatarOptions {
  /// Overrides specific palette entries. Overridden colors bypass the
  /// contrast guarantee.
  final Map<String, String>? palette;

  /// Locks the hue in degrees, so the name drives shape only.
  final double? hue;

  /// Locks the tone as a 0-1 position in the swatch set, pale to ink.
  ///
  /// The swatches are banded with half-open edges, so an exact `1` sits on the
  /// top edge rather than under it and falls back to the first swatch: `1`
  /// renders what `0` renders. Reach for ink with `0.999`.
  final double? tone;

  /// Pins individual traits, so the name drives only what you leave out.
  ///
  /// Each value is the 0-1 position the hash would have produced for that
  /// key — the same units the layout reads. Values outside [0, 1) are
  /// clamped.
  final Map<String, Object>? traits;

  /// Applies NFC + trim + lowercase to the name. Default true.
  final bool normalize;

  /// Enforces the minimum contrast ratios. Default true.
  final bool contrast;

  /// The backdrop the renderer draws behind the figure. Default: the style's
  /// own, which for gen-2 is none.
  final Backdrop? background;

  /// The static pose to apply. Omitted and [idle] are equivalent.
  final Expression? expression;

  const BlobatarOptions({
    this.palette,
    this.hue,
    this.tone,
    this.traits,
    this.normalize = true,
    this.contrast = true,
    this.background,
    this.expression,
  });

  /// Value equality: two option sets render the same avatar iff they are
  /// equal. Maps compare by their entries (values may be a number or a list
  /// of numbers), so a `RepaintBoundary`/`shouldRepaint` decision never
  /// repaints for an unrelated property change.
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! BlobatarOptions) return false;
    return normalize == other.normalize &&
        contrast == other.contrast &&
        hue == other.hue &&
        tone == other.tone &&
        background == other.background &&
        expression == other.expression &&
        _mapEquals(palette, other.palette, _scalarEquals) &&
        _mapEquals(traits, other.traits, _scalarEquals);
  }

  @override
  int get hashCode => Object.hash(
        normalize,
        contrast,
        hue,
        tone,
        background,
        expression,
        Object.hashAllUnordered(palette?.entries ?? const []),
        Object.hashAllUnordered(traits?.entries ?? const []),
      );
}

bool _scalarEquals(Object? a, Object? b) {
  if (a == b) return true;
  if (a is List && b is List) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
  return false;
}

/// Compares two maps by [valueEquals] on each key's values (order agnostic,
/// like the pure-Dart core).
bool _mapEquals(Map<Object?, Object?>? a, Map<Object?, Object?>? b,
    bool Function(Object?, Object?) valueEquals) {
  if (a == null || b == null) return a == b;
  if (a.length != b.length) return false;
  for (final MapEntry<Object?, Object?> entry in a.entries) {
    final Object? other = b[entry.key];
    if (!valueEquals(entry.value, other)) return false;
  }
  return true;
}

/// The resolved inputs: the trait reader and the final palette.
class Resolved {
  final Traits t;
  final Palette palette;
  const Resolved(this.t, this.palette);
}

/// Normalizes and hashes the name once, builds the trait reader, and resolves
/// the palette — including the `hue`/`tone` friendly units and the palette
/// overrides. `hue`/`tone` win over `traits` entries for the same two traits.
Resolved resolve(String seed,
    [BlobatarOptions opts = const BlobatarOptions()]) {
  final Traits t = traitsFor(
    seed,
    normalize: opts.normalize,
    overrides: opts.traits,
  );
  final Map<String, String> p = palette(
    opts.hue ?? t.numIn('hue', 0, 360),
    opts.contrast,
    opts.tone ?? t('tone'),
  );
  if (opts.palette != null) {
    p.addAll(opts.palette!);
  }
  return Resolved(t, p);
}

/// The plate behind the figure, as geometry rather than as markup.
class BackdropGeometry {
  final BlobPath path;
  final String fill;
  const BackdropGeometry(this.path, this.fill);
}

/// The backdrop is the style's concern to default, not the renderer's.
BackdropGeometry? backdropFor(Backdrop? background, Palette p,
    {Backdrop styleDefault = Backdrop.none}) {
  final Backdrop bg = background ?? styleDefault;
  switch (bg) {
    case Backdrop.none:
      return null;
    case Backdrop.square:
      // `M0 0H100V100H0Z` — box(50, 50, 50, 50) emits exactly this.
      return BackdropGeometry(box(50, 50, 50, 50), p[colorBg]!);
    case Backdrop.circle:
      return BackdropGeometry(
        superellipse(const Superellipse(cx: 50, cy: 50, rx: 50, ry: 50, n: 2)),
        p[colorBg]!,
      );
    case Backdrop.squircle:
      return BackdropGeometry(
        superellipse(const Superellipse(cx: 50, cy: 50, rx: 50, ry: 50, n: 6)),
        p[colorBg]!,
      );
  }
}

/// The blobatar in the pieces a renderer needs.
///
/// The Dart analogue of `_layout` at 2.4.0 (without the pose half, which is a
/// later phase): the resolved palette next to the full numeric layout, so
/// color assertions can read the same numbers a renderer paints.
BlobatarLayout layoutFor(String name,
    [BlobatarOptions opts = const BlobatarOptions()]) {
  final Resolved r = resolve(name, opts);
  final BlobatarLayout layout = style.layout(r.t);
  return opts.expression == null
      ? layout
      : bakePose(layout, opts.expression!.pose);
}

/// Layout and palette together, for renderers that own the drawing.
(BlobatarLayout, Palette) partsFor(String name,
    [BlobatarOptions opts = const BlobatarOptions()]) {
  final Resolved r = resolve(name, opts);
  final Expression? expression = opts.expression;
  final BlobatarLayout layout = style.layout(r.t);
  if (expression == null) return (layout, r.palette);
  return (
    bakePose(layout, expression.pose),
    expressionPalette(r.palette, expression),
  );
}
